src/client_side.cc

```diff
@@ -3025,10 +3048,14 @@ ConnStateData::parseTlsHandshake()
         Must(context && context->http);
         HttpRequest::Pointer request = context->http->request;
         debugs(83, 5, "Got something other than TLS Client Hello. Cannot SslBump.");
-        sslBumpMode = Ssl::bumpSplice;
-        context->http->al->ssl.bumpMode = Ssl::bumpSplice;
-        if (!clientTunnelOnError(this, context, request, HttpRequestMethod(), ERR_PROTOCOL_UNKNOWN))
+        if (!clientTunnelOnError(this, context, request, HttpRequestMethod(), ERR_PROTOCOL_UNKNOWN)) {
+            // XXX: The client did not necessarily abort (and may not be using TLS)
+            request->detailError(ERR_PROTOCOL_UNKNOWN, ERR_DETAIL_TLS_HANDSHAKE_ABORTED);
```

Comment:
I suspect that tunnelOnError() should detail its own errors (before returning false) so that we do not have to use a misleading `ERR_DETAIL_TLS_HANDSHAKE_ABORTED` for every tunneling failure. Same for initiateTunneledRequest() (and, hence, splice()).

AFAICT, tunnelOnError() and initiateTunneledRequest() error details are:
* `ERR_DETAIL_CONFIGURATION`: tunneling denied by configuration (including the default setting)
* `ERR_DETAIL_BUFFER`: tunneling impossible because we did not preserve all the client payload
* `ERR_DETAIL_TARGET_ADDRESS`: tunneling impossible because we cannot construct the target URL

Reply:
Mixed feelings. 
 > ERR_DETAIL_CONFIGURATION: tunneling denied by configuration (including the default setting)

In this case we should return the original error details which caused the tunnelOnError to be called.

> ERR_DETAIL_BUFFER: tunneling impossible because we did not preserve all the client payload

This is looks like a bug, however we may have some cases which are not handled correctly by squid, this is may help us to find such cases.
I add this error detail code (commit 36861d).

> ERR_DETAIL_TARGET_ADDRESS: tunneling impossible because we cannot construct the target URL

This is can caused only because of a squid bug. In all cases we are allowing tunneling after an error (transparent proxying and normal CONNECT requests) we are able to compute the final destination. My current code throws a TextException in this case and logs an ExceptionErrorDetail.